src/lexer.cpp

```diff
@@ -412,16 +343,18 @@ TokenArray tokenize(char const* p) {
 		char const* const p0 = p;
 
 		int state = state_count - 1;
-		while (MainLexer::EndStates::Count < state) {
-			state = a.go(state, *p++);
+		int new_state = a.go(state, *p++);
+		while (new_state != MainLexer::EndStates::Error) {
+			state = new_state;
+			new_state = a.go(state, *p++);
 		}
 
-		if (state == MainLexer::EndStates::Error) {
+		if (MainLexer::EndStates::Count <= state) {
 			print_error(p);
 			break;
 		}
```

Comment:
I'm struggling with this logic, tell me if i understood correctly:

since we always stop before reaching the error state, an error manifests as stopping in a non-end state?

Reply:
We stop when the next step is an error state. So if we are not in an end state when that happens, then the thing is clearly wrong. So yes, you got that right.